File: tests/fixtures/grammar/constructs.go
package grammar

// Every control-flow construct in the Go specification's Statements
// section that can carry a decision. One function per construct so a
// disagreement names the construct rather than a file.

func ifElse(v int) int {
	if v > 0 {
		return 1
	} else if v < 0 {
		return -1
	}
	return 0
}

func forThreeClause(n int) int {
	total := 0
	for i := 0; i < n; i++ {
		total += i
	}
	return total
}

func forRange(items []int) int {
	total := 0
	for _, item := range items {
		total += item
	}
	return total
}

func forCondition(n int) int {
	for n > 0 {
		n--
	}
	return n
}

func expressionSwitch(v int) string {
	switch v {
	case 1:
		return "one"
	case 2:
		return "two"
	default:
		return "many"
	}
}

func typeSwitch(v interface{}) string {
	switch v.(type) {
	case int:
		return "int"
	case string:
		return "string"
	}
	return "other"
}

func selectStatement(a chan int, b chan int) int {
	select {
	case x := <-a:
		return x
	case y := <-b:
		return y
	}
}

func booleanOperators(a bool, b bool, c bool) bool {
	return a && b || c
}

func gotoStatement(n int) int {
retry:
	if n > 0 {
		n--
		goto retry
	}
	return n
}

func (b *box[T]) genericMethod(v T) T {
	if b.empty {
		return v
	}
	return b.item
}

File: tests/fixtures/grammar/constructs.f90
! Every control-flow construct in the Fortran standard that can carry a
! decision. One procedure per construct.

subroutine if_else(v, r)
  integer, intent(in) :: v
  integer, intent(out) :: r
  if (v > 0) then
     r = 1
  else if (v < 0) then
     r = -1
  else
     r = 0
  end if
end subroutine if_else

subroutine do_loop(n, total)
  integer, intent(in) :: n
  integer, intent(out) :: total
  integer :: i
  total = 0
  do i = 1, n
     total = total + i
  end do
end subroutine do_loop

subroutine do_while_loop(n, r)
  integer, intent(in) :: n
  integer, intent(out) :: r
  r = n
  do while (r > 0)
     r = r - 1
  end do
end subroutine do_while_loop

subroutine select_case(v, r)
  integer, intent(in) :: v
  integer, intent(out) :: r
  select case (v)
  case (1)
     r = 1
  case (2)
     r = 2
  case default
     r = 0
  end select
end subroutine select_case

subroutine word_operators(a, b, r)
  logical, intent(in) :: a, b
  logical, intent(out) :: r
  r = a .and. b .or. .not. a
end subroutine word_operators

subroutine where_construct(a)
  real, intent(inout) :: a(:)
  where (a > 0.0)
     a = a * 2.0
  end where
end subroutine where_construct
